File: P160 Intersection of Two Linked Lists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        return getIntersectionNode1(headA, headB);
    }
    
    // Approach 1: Hash Table
    // Time: O(n + m), where n is the length of linked list A and m is the length of B
    // Space: o(m)
    private ListNode getIntersectionNode1(ListNode headA, ListNode headB) {
        Set<ListNode> visitedNodes = new HashSet<>();
        while (headB != null) {
            visitedNodes.add(headB);
            headB = headB.next;
        }
        
        while (headA != null) {
            if (visitedNodes.contains(headA)) {
                return headA;
            }
            headA = headA.next;
        }
        
        return null;
    }
    
    // Approach 2: Two pointers
    // Time: O(n + m)
    // Space: o(1)
    private ListNode getIntersectionNode2(ListNode headA, ListNode headB) {
        // We need to know the length of each linked list
        // If one of the pointers reaches the end, just go back to the beginning and eventually both pointers will meet at the intersection node. 
        // If there's no intersection between two linked lists, two pointers will be the end (null) at the same time
        ListNode p1 = headA;
        ListNode p2 = headB;
        while (p1 != p2) {
            // faster if switching the heads, when the pointers reach the end
            p1 = p1 == null ? headB : p1.next; 
            p2 = p2 == null ? headA : p2.next;
        }
        return p1;
    }
}

File: P160 Intersection of Two Linked Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        return self._getIntersectionNode2(headA, headB)
    
    # Approach 1: Hash Table
    # Time: O(n + m), where n is the length of linked list A and m is the length of B
    # Space: o(m)
    def _getIntersectionNode1(self, headA: ListNode, headB: ListNode) -> ListNode:
        visited_nodes = set()
        while headB:
            visited_nodes.add(headB)
            headB = headB.next
            
        while headA:
            if headA in visited_nodes:
                return headA
            headA = headA.next
        return None
    
    # Approach 2: Two pointers
    # Time: O(n + m)
    # Space: o(1)
    def _getIntersectionNode2(self, headA: ListNode, headB: ListNode) -> ListNode:
        p1, p2 = headA, headB
        while p1 != p2:
            p1 = p1.next if p1 else headB
            p2 = p2.next if p2 else headA
        return p1